From 0a5ad039c5b4222655aa434e8956e8e7ad6df5c2 Mon Sep 17 00:00:00 2001
From: Aurelio Colosimo <aurelio@aureliocolosimo.it>
Date: Tue, 10 Mar 2015 16:32:32 +0100
Subject: [PATCH 195/198] AR933X uart driver: handle TX_ORIDE and RX_ORIDE
 registers
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This patch is mostly needed for AR934X chip, but is anyway correct
for AR933X too: when hw flow control is disabled, UART_TX_READY_ORIDE
and UART_RX_READY_ORIDE bits of UART_CS register must be set
to 1 (see ยง7.2.2 of AR9331 datasheet).

Signed-off-by: Aurelio Colosimo <aurelio@aureliocolosimo.it>
---
 drivers/tty/serial/ar933x_uart.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/tty/serial/ar933x_uart.c b/drivers/tty/serial/ar933x_uart.c
index c586481..dba33bb 100644
--- a/drivers/tty/serial/ar933x_uart.c
+++ b/drivers/tty/serial/ar933x_uart.c
@@ -302,6 +302,14 @@ static void ar933x_uart_set_termios(struct uart_port *port,
 	ar933x_uart_rmw_set(up, AR933X_UART_CS_REG,
 			    AR933X_UART_CS_HOST_INT_EN);
 
+	/* enable TX ready override */
+	ar933x_uart_rmw_set(up, AR933X_UART_CS_REG,
+			AR933X_UART_CS_TX_READY_ORIDE);
+
+	/* enable RX ready override */
+	ar933x_uart_rmw_set(up, AR933X_UART_CS_REG,
+			AR933X_UART_CS_RX_READY_ORIDE);
+
 	/* reenable the UART */
 	ar933x_uart_rmw(up, AR933X_UART_CS_REG,
 		        AR933X_UART_CS_IF_MODE_M << AR933X_UART_CS_IF_MODE_S,
@@ -440,6 +448,10 @@ static int ar933x_uart_startup(struct uart_port *port)
 	ar933x_uart_rmw_set(up, AR933X_UART_CS_REG,
 			    AR933X_UART_CS_HOST_INT_EN);
 
+	/* enable TX ready override */
+	ar933x_uart_rmw_set(up, AR933X_UART_CS_REG,
+			AR933X_UART_CS_TX_READY_ORIDE);
+
 	/* Enable RX interrupts */
 	up->ier = AR933X_UART_INT_RX_VALID;
 	ar933x_uart_write(up, AR933X_UART_INT_EN_REG, up->ier);
-- 
1.7.9.5

